### PR TITLE
fix(Python): Support orphaned shapes, update Config shape generation

### PR DESCRIPTION
**Pull request by lucasmcdonald3**
*Issue #, if available:*

*Description of changes:*

Python:
* Supported "orphaned" shapes (shapes with no graph connection to the LocalService)
* Refactor LocalService Config shape generation from its own logic to "normal" shape generation
  * Before, Config shapes were "orphaned" and could not be discovered by normal shape generation. I wrote manual generation for Config shapes.
  * Now, Config shape generation is more closely tied to normal shape generation, which gets it more fully-featured shape generation (constraints, as/from_dict)
* Don't use forward references for boto3 client typehints

Check out this MPL PR for sample generated code: https://github.com/aws/aws-cryptographic-material-providers-library/pull/831

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: codegen/smithy-dafny-codegen-modules/smithy-python/codegen/smithy-python-codegen/src/main/java/software/amazon/smithy/python/codegen/StructureGenerator.java
Comment (lucasmcdonald3) [inline codegen/smithy-dafny-codegen-modules/smithy-python/codegen/smithy-python-codegen/src/main/java/software/amazon/smithy/python/codegen/StructureGenerator.java:255]: Unfortunately, this change means another

`make mvn_local_deploy_polymorph_python_dependencies`

on merging this from main, but I think this is better than duplicating the code.

Comment (robin-aws) [reply codegen/smithy-dafny-codegen-modules/smithy-python/codegen/smithy-python-codegen/src/main/java/software/amazon/smithy/python/codegen/StructureGenerator.java:255]: I'm inclined to add `mvn_local_deploy_polymorph_dependencies` as a dependency of all polymorph targets, given it should be cheap and hopefully Gradle-cached anyway.

File: codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/utils/ModelUtils.java
Comment (robin-aws) [inline codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/utils/ModelUtils.java:735]: Nice that this is already available for you!

File: codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/utils/ModelUtils.java
Comment (robin-aws) [inline codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/utils/ModelUtils.java]:
```suggestion
        ModelUtils.isInServiceNamespace(shape, serviceShape) &&
```

Comment (robin-aws) [reply codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/utils/ModelUtils.java]: And likewise below. Just reads a bit better.

File: codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/utils/ModelUtils.java
Comment (robin-aws) [inline codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/utils/ModelUtils.java:749]: I wasn't 100% sure from the TopologicalIndex documentation, but is this guaranteed to be a superset of `getOrderedShapes()`?

Would be nice to have an explicit unit test just for `getTopologicallyOrderedOrphanedShapesForService`, especially since it can clarify what is or isn't a recursive shape.

**Review by robin-aws — CHANGES_REQUESTED** [review]
Quite happy with the general approach to handle orphaned shapes. Cut https://github.com/smithy-lang/smithy-dafny/issues/627 for longer term follow up on the problem.

Are you dependent on an OrphanedShapes test model to get added to test these changes?

File: codegen/smithy-dafny-codegen-modules/smithy-python/codegen/smithy-python-codegen/src/main/java/software/amazon/smithy/python/codegen/StructureGenerator.java
Comment (lucasmcdonald3) [reply codegen/smithy-dafny-codegen-modules/smithy-python/codegen/smithy-python-codegen/src/main/java/software/amazon/smithy/python/codegen/StructureGenerator.java:255]: Good idea. I'll include this

File: codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/utils/ModelUtils.java
Comment (lucasmcdonald3) [reply codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/utils/ModelUtils.java:749]: > I wasn't 100% sure from the TopologicalIndex documentation, but is this guaranteed to be a superset of getOrderedShapes()?

I suspect they're disjoint, otherwise some shapes would be generated multiple times.

> Would be nice to have an explicit unit test just for getTopologicallyOrderedOrphanedShapesForService, especially since it can clarify what is or isn't a recursive shape.

I think a unit test might be more effort than it's worth --

Primarily, I'm inclined to trust whatever Smithy-Core is doing here, but verify it against a TestModel/real projects. I built this PR against the MPL and DBESDK, it picked up orphaned shapes, and I can run the code. I don't think the test you're proposing gets us as much ROI.

For a unit test, we'd have to create a Smithy model as a POJO (sort of like [this](https://github.com/smithy-lang/smithy-dafny/blob/main-1.x/codegen/smithy-dafny-codegen/src/test/java/software/amazon/polymorph/util/TestModel.java)), which seems more brittle than is worth doing.

Let me know your thoughts.

**Comment by lucasmcdonald3** [comment]
> Are you dependent on an OrphanedShapes test model to get added to test these changes?

I don't think so, but let me know if you disagree.

I built this PR into the MPL and DBESDK, and both of those projects are still passing.
* MPL PR: https://github.com/aws/aws-cryptographic-material-providers-library/pull/831
* DBESDK PR doesn't exist. This project only exists on my local right now.

This (plus the MPL PR) is ideally the last PR I'll to land before I release the Python MPL, so I'm trying to push any non-essential work out until after that lands.
I'm carving out some time to handle these post-launch fast follows, and I added OrphanedShapes TestModel to that list

File: codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/utils/ModelUtils.java
Comment (robin-aws) [reply codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/utils/ModelUtils.java:749]: > I suspect they're disjoint, otherwise some shapes would be generated multiple times.

Good point, I think they must be.

> For a unit test, we'd have to create a Smithy model as a POJO (sort of like [this](https://github.com/smithy-lang/smithy-dafny/blob/main-1.x/codegen/smithy-dafny-codegen/src/test/java/software/amazon/polymorph/util/TestModel.java)), which seems more brittle than is worth doing.

You can include a *.smithy file as a resource and read it in pretty easily instead: https://github.com/smithy-lang/smithy-dafny/blob/main-1.x/codegen/smithy-dafny-codegen/src/test/java/software/amazon/polymorph/smithyjava/modeled/ModeledShapeValueTest.java#L77-L82

But I'll leave it to your judgement on whether you want to bother with that on this PR, I won't block on it.

Comment (lucasmcdonald3) [reply codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/utils/ModelUtils.java:749]: That's pretty nifty -- I'll plan to do this when I do the OrphanedShapes TestModel, but not now.

File: SmithyDafnyMakefile.mk
Comment (robin-aws) [inline SmithyDafnyMakefile.mk:275]: ❤️

File: codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithypython/localservice/extensions/DafnyPythonLocalServiceStructureGenerator.java
Comment (robin-aws) [inline codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithypython/localservice/extensions/DafnyPythonLocalServiceStructureGenerator.java:108]: Was this lifted from smithy-python too then? It's a great idea but I wasn't sure if they'd acted on it yet.

File: codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithypython/localservice/extensions/DafnyPythonLocalServiceStructureGenerator.java
Comment (robin-aws) [inline codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithypython/localservice/extensions/DafnyPythonLocalServiceStructureGenerator.java:313]: Nit picking, but isn't it more accurately that shapes from dependencies are already declared?

File: codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithypython/wrappedlocalservice/extensions/DafnyPythonWrappedLocalServiceClientCodegenPlugin.java
Comment (robin-aws) [inline codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithypython/wrappedlocalservice/extensions/DafnyPythonWrappedLocalServiceClientCodegenPlugin.java:58]: off topic question: what does this transformation do exactly?

**Review by robin-aws — APPROVED** [review]
LGTM, some non-blocking questions

File: codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithypython/wrappedlocalservice/extensions/DafnyPythonWrappedLocalServiceClientCodegenPlugin.java
Comment (lucasmcdonald3) [reply codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithypython/wrappedlocalservice/extensions/DafnyPythonWrappedLocalServiceClientCodegenPlugin.java:58]: Check out https://github.com/smithy-lang/smithy-dafny/blob/main-1.x/codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithypython/wrappedlocalservice/extensions/DafnyPythonWrappedLocalServiceClientCodegenPlugin.java#L52

In preprocessing, it replaces a LocalServiceTrait on the ServiceTrait with a WrappedLocalServiceTrait.
This results in some different codegen; wrapped LocalServices don't get all the codegen from LocalServices. (No models, etc.)

File: codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithypython/localservice/extensions/DafnyPythonLocalServiceStructureGenerator.java
Comment (lucasmcdonald3) [reply codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithypython/localservice/extensions/DafnyPythonLocalServiceStructureGenerator.java:108]: Yes

File: codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithypython/localservice/extensions/DafnyPythonLocalServiceStructureGenerator.java
Comment (lucasmcdonald3) [reply codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithypython/localservice/extensions/DafnyPythonLocalServiceStructureGenerator.java:313]: This is actually about circular dependencies within one project.
I'm going to write about this problem and solutions in a Smithy-Dafny Python dev guide, but an overview/instance of the problem is that `models.py` can't import `references.py` at the top-level because `references.py` depends on `models.py` at the top-level, but `models.py` still needs to import `references.py` to use its classes.
One solution is to defer the imports like I've done here.
I wrote a fair amount about this and drew some graphs, but I'm going to collate all that later.

File: codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithypython/wrappedlocalservice/extensions/DafnyPythonWrappedLocalServiceClientCodegenPlugin.java
Comment (robin-aws) [reply codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithypython/wrappedlocalservice/extensions/DafnyPythonWrappedLocalServiceClientCodegenPlugin.java:58]: Ah thanks! I wasn't aware of that trait.

File: codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithypython/localservice/extensions/DafnyPythonLocalServiceStructureGenerator.java
Comment (robin-aws) [reply codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithypython/localservice/extensions/DafnyPythonLocalServiceStructureGenerator.java:313]: I will be thrilled to review a PR that targets `docs/python` in the future :)

**Pull request merged**
